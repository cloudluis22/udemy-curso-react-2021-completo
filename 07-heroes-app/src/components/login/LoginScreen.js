import React, { useContext } from 'react';
import AuthContext from '../../auth/AuthContext';
import types from '../../types/types';

export const LoginScreen = ({ history }) => {

    const { dispatch } = useContext(AuthContext)

    const handleLogin = () => {

       const lastPath = localStorage.getItem('lastPath') || '/';

       const newUser = {
            name: 'Sesi√≥n Iniciada',
            logged: true
        }

        const action = {
            type: types.login,
            payload: newUser
        }

        dispatch(action);

        history.replace(lastPath);

    };

    return (
        <div className="container mt-5">

            <h1>Login Screen</h1>
            <hr />

            <button
                className="btn btn-primary"
                onClick={ handleLogin }
            >
                Login
            </button>

        </div>
    )
}
