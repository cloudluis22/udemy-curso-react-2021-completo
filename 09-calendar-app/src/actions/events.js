import { __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED } from "react-dom";
import Swal from "sweetalert2";
import { convertEvents } from "../helpers/convertEvents";
import { fetchToken } from "../helpers/fetch";
import { types } from "../types/types";

export const eventStartAddNew = (event) => {
    return async ( dispatch, getState ) => {

        const { uid, name } = getState().auth;

        try {
            const dbEvent = {
                ...event,
                start: event.startDate,
                end: event.endDate
            }

            const response = await fetchToken('events', dbEvent, 'POST');
            const body = await response.json();

            if (body.ok) {
                event.id = body.event.id
                event.user = {
                    _id: uid,
                    name: name
                }
                console.log(event);
                dispatch( eventAddNew(event) );
            }

        } catch (error) {
            console.log(error)
        }

    };
};

export const eventStartLoading = () => {
    return async (dispatch) => {

        try {
            const response = await fetchToken('events');
            const body = await response.json();

            const events = convertEvents(body.events);
           dispatch( eventsLoaded( events ) );

        } catch (error) {

        }

    };
};

export const eventStartUpdate = (event) => {
    return async (dispatch) => {

        try {

            const response = await fetchToken(`events/${event.id}`, event, 'DELETE');
            const body = await response.json();

            if (body.ok) {
                dispatch( eventUpdated(event) );
            }
            else {
                Swal.fire('Error', body.msg, 'error');
            }

        } catch (error) {
            console.log(error);
        };

    };
};

export const eventStartDelete = () => {


    return async (dispatch, getState) => {
        try {

            const { id } = getState().calendar.activeEvent;

            const response = await fetchToken(`events/${id}`, {}, 'delete');
            const body = await response.json();

            if (body.ok) {
                dispatch( eventDeleted() );
            }
            else {
                Swal.fire('Error', body.msg, 'error');
            }

        } catch (error) {
            console.log(error);
        };

    };
}

export const eventLogoutCleaning = () => ({ type: types.eventLogoutCleaning });

const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
});

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActive = () => ({
    type: types.eventClearActive
});

const eventUpdated = ( event ) => ({
    type: types.eventUpdated,
    payload: event
});

const eventDeleted = () => ({
    type: types.eventDeleted
});

const eventsLoaded = (events) => ({ type: types.eventsLoaded, payload: events });