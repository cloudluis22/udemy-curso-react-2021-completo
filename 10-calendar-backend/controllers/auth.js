const { response } = require('express');
const bcrypt = require('bcryptjs');
const User = require('../models/User-model');
const { generateJWT } = require('../helpers/jwt');

const userCreate = async (request, response = response) => {
  const { email, password } = request.body;

  try {
    let user = await User.findOne({ email });

    if (user) {
      return response.status(400).json({
        ok: false,
        msg: 'E-mail already registered with another user.',
      });
    }

    user = new User(request.body);

    // Encriptar password.
    const salt = bcrypt.genSaltSync();
    user.password = bcrypt.hashSync(password, salt);

    await user.save();

    // Generar JWT.
    const token = await generateJWT(user.id, user.name);

    response.status(201).json({
      ok: true,
      uid: user.id,
      name: user.name,
      token,
    });
  } catch (error) {
    console.log(error);
    response.status(500).json({
      ok: false,
      msg: 'Unexpected error, please contact the admin.',
    });
  }
};

const userLogin = async (request, response = response) => {
  const { email, password } = request.body;

  try {
    const user = await User.findOne({ email });

    if (!user) {
      return response.status(400).json({
        ok: false,
        msg: 'User not found with given e-mail.',
      });
    }

    // Verificar la contraseÃ±a.
    const validPassword = bcrypt.compareSync(password, user.password);
    if (!validPassword) {
      return response.status(400).json({
        ok: false,
        msg: 'Incorrect password',
      });
    }

    const token = await generateJWT(user.id, user.name);

    response.json({
      ok: true,
      uid: user.id,
      name: user.name,
      token,
    });
  } catch (error) {
    console.log(error);
    response.status(500).json({
      ok: false,
      msg: 'Unexpected error, please contact the admin.',
    });
  }
};

const tokenRevalidate = async (request, response = response) => {
  const { uid, name } = request;

  // Generar un nuevo JWT.
  const token = await generateJWT(uid, name);

  response.json({
    ok: true,
    token,
    uid,
    name,
  });
};

module.exports = {
  userCreate,
  userLogin,
  tokenRevalidate,
};
