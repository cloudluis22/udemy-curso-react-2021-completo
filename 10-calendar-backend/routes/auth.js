/*
    Auth Routes
    /api/auth
*/

const { Router } = require('express');
const { check } = require('express-validator');

const { userCreate, userLogin, tokenRevalidate } = require('../controllers/auth');
const { validateFields } = require('../middlewares/validate-fields');
const { validateJWT } = require('../middlewares/validate-jwt');

const router = Router();

router.post(
    '/new',
    [
        check('name', 'Name is required.').not().isEmpty(),
        check('email', 'Incorrect e-mail / not included.').isEmail(),
        check('password', 'Password must be at least 6 characters.').isLength({ min: 6 }),
        validateFields
    ],
    userCreate
);

router.post(
    '/',
    [
        check('email', 'Incorrect e-mail / not included.').isEmail(),
        check('password', 'Password must be at least 6 characters').isLength({ min: 6 }),
        validateFields
    ],
    userLogin
);

router.get('/renew', validateJWT, tokenRevalidate);

module.exports = router;