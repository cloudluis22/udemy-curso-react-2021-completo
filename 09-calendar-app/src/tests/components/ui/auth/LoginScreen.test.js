import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import Swal from 'sweetalert2';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Enzyme from 'enzyme';

import { LoginScreen } from '../../../../components/auth/LoginScreen';

import '@testing-library/jest-dom';
import { startLogin, startRegister } from '../../../../actions/auth';


jest.mock('../../../../actions/auth', () => ({
    startLogin: jest.fn(),
    startRegister: jest.fn()
}));

jest.mock('sweetalert2', () => ({
    fire: jest.fn()
}));

Enzyme.configure({ adapter: new Adapter() });

const middlewares = [ thunk ];
const mockStore = configureStore(middlewares);

const initState = {};
const store = mockStore(initState);

store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <LoginScreen />
    </Provider>
);

describe('Pruebas en el componente de "<LoginScreen />".', () => {


    beforeEach(() => {
        jest.clearAllMocks();
    })

    test('Debe de mostrarse correctamente.', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de llamar el dispatch de el login.', () => {

        wrapper.find('input[name="lEmail"]').simulate('change', {
            target: {
                name: 'lEmail',
                value: 'bennyuwo@hotmail.com'
            }
        });

        wrapper.find('input[name="lPassword"]').simulate('change', {
            target: {
                name: 'lPassword',
                value: '123456'
            }
        });

        wrapper.find('form').at(0).prop('onSubmit')({
            preventDefault(){}
        });

        expect(startLogin).toHaveBeenCalledWith('bennyuwo@hotmail.com', '123456');

    });

    test('Debe de prohibír el registro si las contraseñas son diferentes.', () => {


        wrapper.find('input[name="rEmail"]').simulate('change', {
            target: {
                name: 'rEmail',
                value: 'test@hotmail.com'
            }
        });

        wrapper.find('input[name="rPassword"]').simulate('change', {
            target: {
                name: 'rPassword',
                value: '123456'
            }
        });

        wrapper.find('input[name="rPassword2"]').simulate('change', {
            target: {
                name: 'rPassword2',
                value: '12345678'
            }
        });

        wrapper.find('form').at(1).prop('onSubmit')({
            preventDefault(){}
        });

        expect(startRegister).not.toHaveBeenCalled();
        expect(Swal.fire).toHaveBeenCalledWith('Error', 'Las contraseñas tienen que coincidir.', 'error');

    });

    test('Debe de funcionar el registro con contraseñas iguales.', () => {

        wrapper.find('input[name="rEmail"]').simulate('change', {
            target: {
                name: 'rEmail',
                value: 'test@hotmail.com'
            }
        });

        wrapper.find('input[name="rPassword"]').simulate('change', {
            target: {
                name: 'rPassword',
                value: '123456'
            }
        });

        wrapper.find('input[name="rPassword2"]').simulate('change', {
            target: {
                name: 'rPassword2',
                value: '123456'
            }
        });

        expect(Swal.fire).not.toHaveBeenCalled();

    });

});