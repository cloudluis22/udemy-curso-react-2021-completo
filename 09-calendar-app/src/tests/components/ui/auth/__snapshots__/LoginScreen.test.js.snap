// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente de "<LoginScreen />". Debe de mostrarse correctamente. 1`] = `ReactWrapper {}`;
